File: src/main/python/net/coderextreme/data/CloudsProcedural4.py
'''
x3d.py package 4.0.63.7 loaded, have fun with X3D Graphics!
'''
from x3d import *
from x3d import SFBool
print(
X3D(profile="Immersive", version="3.2", 
head=head(), 
Scene=Scene(children=[
#A png image file for the cloud texture must be designated in the ecmascript node.

Viewpoint(description="Main", jump=False, orientation=((0,1,0,1.57)), position=((50000,1000,42000))),
Viewpoint(description="Light House Tower", jump=False, orientation=((0,1,0,1.3)), position=((45000,1290,44000))),
Viewpoint(description="centerWest", jump=False, orientation=((0,1,0,2.5)), position=((48000,1000,20000))),
Background(groundColor=[(0,0,1)], skyColor=[(0,0,1)]),
DirectionalLight(ambientIntensity=1, direction=((-1,0,0)), global_=True),
Group(DEF="Terrain", children=[
Transform(scale=((50,50,50)), translation=((25000,0,25000)), children=[
Inline(url=["MontereyBayLargeMesh.x3d","https://savage.nps.edu/Savage/Environment/Atmosphere/MontereyBayLargeMesh.x3d","MontereyBayLargeMesh.wrl","https://savage.nps.edu/Savage/Environment/Atmosphere/MontereyBayLargeMesh.wrl"])]),
Transform(rotation=((1,0,0,1.57)), translation=((25000,0,25000)), children=[
Shape(
geometry=
Rectangle2D(size=((77000,55000))), 
appearance=
Appearance(
texture=
ImageTexture(url=["ocean.png","https://savage.nps.edu/Savage/Environment/Atmosphere/ocean.png"])))])]),
Group(DEF="Placemarks", children=[
Transform(scale=((50,50,50)), translation=((45000,30,44000)), children=[
Inline(url=["Lighthouse.x3d","https://savage.nps.edu/Savage/Environment/Atmosphere/Lighthouse.x3d","Lighthouse.wrl","https://savage.nps.edu/Savage/Environment/Atmosphere/Lighthouse.wrl"])])]),
Group(DEF="Clouds", children=[
Transform(DEF="Cumulus"),
Transform(DEF="Cirrus"),
Transform(DEF="Fog"),
Script(DEF="PixelScript", directOutput=True, field=[field(name="Cumulus", accessType="initializeOnly", type="SFNode", children=[
Transform(USE="Cumulus")]),
field(name="Cirrus", accessType="initializeOnly", type="SFNode", children=[
Transform(USE="Cirrus")]),
field(name="Fog", accessType="initializeOnly", type="SFNode")
], 
#['', '', 'ecmascript:', '', '', 'function cumulustranslation() // These values designate the boundary location of the cloud', '{', '\tX = 50000*Math.random();          //  X horizontal range', '\tY = 1000 + 300*Math.random();\t //  Y vertical base + range', '\tZ = 50000*Math.random();         // z horizontal range', '', "\trandomt = new String(X+' '+Y+' '+Z);", '', '\treturn randomt;', '\t', '}', '', '', '', 'function cumulusscale() // these values scale a cloud within a designated size', '{', '', '\tvar maxscale = 1;', '', '\tscale = Math.round(9+maxscale*Math.random());', '\tX = 1.5*scale;', '\tY = scale;', '\tZ = scale;', '', "\trandomscale = new String(X+' '+Y+' '+Z);", '', '\treturn randomscale;', '\t', '}', '', '', 'function cirrustranslation() // These values designate the boundary location of the cloud', '{', '\tX = 50000*Math.random();          //  X horizontal range', '\tY = 8000 + 1000*Math.random();\t //  Y vertical base + range', '\tZ = 50000*Math.random();         // z horizontal range', '', "\trandomt = new String(X+' '+Y+' '+Z);", '', '\treturn randomt;', '\t', '}', '', '', '', 'function cirrusscale() // these values scale a cloud within a designated size', '{', '', '\tvar maxscale = 1;', '', '\tscale = Math.round(9+maxscale*Math.random());', '\tX = 1.5*scale;', '\tY = 2*Math.random();', '\tZ = 1.5*scale;', '', "\trandomscale = new String(X+' '+Y+' '+Z);", '', '\treturn randomscale;', '\t', '}', '', '', 'function cumulussectiontranslation() // These random values place another portion of cumulus type cloud', '{', '', '\trandomtheta = 6.28319*Math.random();', '\trandomphi = .7854*Math.random();', '\trandomradius = 90 + 5*Math.random();//the first whole number should be close to the sectionradius', '', '\tX = randomradius*Math.cos(randomtheta)*Math.sin(randomphi);', '\tZ = randomradius*Math.sin(randomtheta)*Math.sin(randomphi);', '\tY = randomradius*Math.cos(randomphi);', '\t', '', "\trandomt = new String(X+' '+Y+' '+Z);", '', '\treturn randomt;', '\t', '}', '', 'function cirrussectiontranslation() // These random values place another portion of cirrus type cloud', '{', '', '\trandomtheta = 6.28319*Math.random();', '\trandomphi = .7854*Math.random();', '\trandomradius = 90 + 5*Math.random();//the first whole number should be close to the sectionradius', '', '\tX = randomradius*Math.cos(randomtheta)*Math.sin(randomphi);', '\tZ = randomradius*Math.sin(randomtheta)*Math.sin(randomphi);', '\tY = randomradius*Math.cos(randomphi);', '\t', '', "\trandomt = new String(X+' '+Y+' '+Z);", '', '\treturn randomt;', '\t', '}', '', '', 'function rotation() // This random value is for the billboard rotation not used in this script', '{', '', '', '\tradians = 6.28*Math.random();', '', "\trandomr = new String('0 0 1 ' + radians );", '\t', '\t', '\treturn randomr;', '\t', '}', '', 'function cumulus()', '{', '', 'var maxi = 20;  // number of clouds', '', 'var maxj = 5; // denotes how many portions affecting the size of the cloud', '', 'var maxk = 8;  // number of billboards indicating cloud density', '', 'var sectionradius = 100;  //radius of individual cloud sections', '', '', '', '', '', 'for (var i=0; i < maxi; i++) ', '{', '', '', '', "CloudStringA = '\tTransform {\t\t
#' +", "'    scale '+ cumulusscale() + '               \t
#' +", "'    translation '+ cumulustranslation() + '    
#' +    // cloud placement", "'    children [\t                                
#';", '', '', 'CloudStringB = new Array();', 'CloudStringF = new Array();', '', '   \tfor (var j=0; j < maxj; j++)', '   \t{', '', '\tradius = 0;', '', "\tCloudStringB[j]= '  Transform {\t\t    \t       
#' +", "\t'    translation '+ cumulussectiontranslation() + '    
#' +     // section placement", "\t'    children [\t                                       
#';", '', '\t', '\tCloudStringC = new Array();', '\timage = new String();', '', '      \t\tfor (var k=1; k < maxk; k++)  // maxk value denotes how many textured billboards make up the cloud ', '      \t\t{', '', '', '\t\trandomtheta = 6.28319*Math.random();', '\t\trandomphi = 1.57079*Math.random();', '\t\tradius = radius+(sectionradius/maxk); // radius incremental steps based on billow radius and max billboards', '', '\t\tX = radius*Math.cos(randomtheta)*Math.sin(randomphi);', '\t\tZ = radius*Math.sin(randomtheta)*Math.sin(randomphi);', '\t\tY = radius*Math.cos(randomphi);', '', '', '\t\tif (Y <= 30) //cloud shading and lighting control', '  \t{\t', '\timage = \' "CloudTexture1_5.png" "https://savage.nps.edu/Savage/Environment/Spheretexture.png" 
#\';', '  \t}', '', '  \t\telse', '  \t{\t', '\timage = \' "CloudTexture1_4.png" "https://savage.nps.edu/Savage/Environment/Spheretexture.png" 
#\';', '  \t}', '', '\t', '\t\t', "\t\tBillboardtranslation = new String(X+' '+Y+' '+Z);", '', "\t\tCloudStringC[k] = '\tTransform {\t\t                
#' +", "\t\t'            translation '+ Billboardtranslation   + '          
#' +     // random billboard placement within radius designated above", "\t\t'\t  children [\t                                        
#' +", "\t\t'\t      Billboard {\t                                
#' +", "\t\t'\t        axisOfRotation 0 0 0\t                        
#' +     // 0 0 0 designates rotation on all axis", "\t\t'\t        children [\t                                
#' +", "\t\t'\t            Transform {\t                \t\t
#' +", "\t\t'\t              rotation  0 0 0 0 \t\t        
#' +     // a rotation of the individual billboards can be defined", "\t\t'\t              children [\t                        
#' +", "\t\t'\t                  Shape {\t                        
#' +", "\t\t'\t                    appearance Appearance {\t        
#' +", "\t\t'\t\t\t\tmaterial Material {\t\t
#' +", "\t\t'\t\t\t\t                }  \t\t
#' +", "\t\t'\t                      texture ImageTexture {\t        
#' +", "\t\t'\t                        url [ ' + image + ' ]           
#' + ", "\t\t'\t                      }\t                                
#' +", "\t\t'\t                    }\t                                
#' +", "\t\t'\t                    geometry IndexedFaceSet {\t        
#' +     // define type of geometry to texture", "\t\t'\t                      coordIndex [ 0, 1, 2, 3 ]\t        
#' +", "\t\t'\t\t\t      solid FALSE\t\t        
#' +", "\t\t'\t                      coord Coordinate {\t        
#' +", "\t\t'\t                        point [ 50 50 0,\t        
#' +     // define size of the geometry. Here 100 meter 2D square.", "\t\t'\t                                50 -50 0,\t        
#' +", "\t\t'\t                               -50 -50 0,\t        
#' +", "\t\t'\t                               -50 50 0 ]\t        
#' +", "\t\t'\t                      }\t                                
#' +", "\t\t'\t                    }\t                                
#' +", "\t\t'\t                  }\t                                
#' +", "\t\t'\t              ]\t                                        
#' +", "\t\t'\t            }\t                                        
#' +", "\t\t'\t       ]\t                                        
#' +", "\t\t'\t   }\t                                                
#' +", "\t\t'      ]\t                                                
#' +", "\t\t'     }\t                                                        
#';      ", '\t\t', '', '\t\t}', '', "\tCloudStringD = CloudStringC.join(' ');", '', '\t', "\tCloudStringE = '   ]\t                 
#' +", "\t'\t}\t                         
#';", '', '\tCloudStringF[j] = CloudStringB[j] + CloudStringD +CloudStringE;', '', '', '\t}', '', "CloudStringG = CloudStringF.join(' ');", '', "CloudStringH = '      ]\t                                        
#' +", "'     }\t                                                        
#' +", "'#########################################################      
#';", '', 'CloudString = CloudStringA + CloudStringG + CloudStringH;', '', '', '', 'newNode = Browser.createVrmlFromString(CloudString);', 'Cumulus.children[i] = newNode[0];', '', '', '   }', '', '}', '', 'function cirrus()', '', '{', '', 'var maxi = 2;  // number of clouds', '', 'var maxj = 5; // denotes how many portions affecting the size of the cloud', '', 'var maxk = 8;  // number of billboards indicating cloud density', '', 'var sectionradius = 1000;  //radius of individual cloud sections', '', '', '', '', '', 'for (var i=0; i < maxi; i++) ', '{', '', '', '', "CloudStringA = '\tTransform {\t\t 
#' +", "'    scale '+ cirrusscale() + '               \t 
#' +", "'    translation '+ cirrustranslation() + '      
#' +    // cloud placement", "'    children [\t                                 
#';", '', '', 'CloudStringB = new Array();', 'CloudStringF = new Array();', '', '   \tfor (var j=0; j < maxj; j++)', '   \t{', '', '\tradius = 0;', '', "\tCloudStringB[j]= '  Transform {\t\t    \t      
#' +", "\t'    translation '+ cirrussectiontranslation() + '    
#' +     // section placement", "\t'    children [\t                                      
#';", '', '\t', '\tCloudStringC = new Array();', '', '      \t\tfor (var k=1; k < maxk; k++)  // maxk value denotes how many textured billboards make up the cloud ', '      \t\t{', '', '', '\t\trandomtheta = 6.28319*Math.random();', '\t\trandomphi = 1.57079*Math.random();', '\t\tradius = radius+(sectionradius/maxk); // radius incremental steps based on section radius and max billboards', '', '\t\tX = radius*Math.cos(randomtheta)*Math.sin(randomphi);', '\t\tZ = radius*Math.sin(randomtheta)*Math.sin(randomphi);', '\t\tY = radius*Math.cos(randomphi);', '\t\t', "\t\tBillboardtranslation = new String(X+' '+Y+' '+Z);", '', "\t\tCloudStringC[k] = '\tTransform {\t\t                
#' +", "\t\t'            translation '+ Billboardtranslation   + '          
#' +     // random billboard placement within radius designated above", "\t\t'\t  children [\t                                        
#' +", "\t\t'\t      Billboard {\t                                
#' +", "\t\t'\t        axisOfRotation 0 0 0\t                        
#' +     // 0 0 0 designates rotation on all axis", "\t\t'\t        children [\t                                
#' +", "\t\t'\t            Transform {\t                \t\t
#' +", "\t\t'\t              rotation '  + rotation() + '\t        
#' +", "\t\t'\t              children [\t                        
#' +", "\t\t'\t                  Shape {\t                        
#' +", "\t\t'\t                    appearance Appearance {\t        
#' +", "\t\t'\t\t\t    material Material {\t\t\t
#' +", "\t\t'\t\t\t    }\t\t\t\t\t
#' +", " \t\t'\t                      texture ImageTexture {\t        
#' +", '\t\t\'\t                        url ["cloudtexture3.png" "https://savage.nps.edu/Savage/Environment/cloudtexture1_4.png" ] 
#\' +', "\t\t'\t                      }\t                                
#' +", "\t\t'\t                    }\t                                
#' +", "\t\t'\t                    geometry IndexedFaceSet {\t        
#' +     // define type of geometry to texture", "\t\t'\t                      coordIndex [ 0, 1, 2, 3 ]\t        
#' +", "\t\t'\t\t\t      solid FALSE\t\t        
#' +", "\t\t'\t                      coord Coordinate {\t        
#' +", "\t\t'\t                        point [ 500 500 0,\t        
#' +     // define size of the geometry. Here 100 meter 2D square.", "\t\t'\t                                500 -500 0,\t        
#' +", "\t\t'\t                               -500 -500 0,\t        
#' +", "\t\t'\t                               -500 500 0 ]\t        
#' +", "\t\t'\t                      }\t                                
#' +", "\t\t'\t                    }\t                                
#' +", "\t\t'\t                  }\t                                
#' +", "\t\t'\t              ]\t                                        
#' +", "\t\t'\t            }\t                                        
#' +", "\t\t'\t       ]\t                                        
#' +", "\t\t'\t   }\t                                                
#' +", "\t\t'      ]\t                                                
#' +", "\t\t'     }\t                                                        
#';      ", '\t\t', '', '\t\t}', '', "\tCloudStringD = CloudStringC.join(' ');", '', "\tCloudStringE = '   ]\t                 
#' +", "\t'\t}\t                         
#';", '', '\tCloudStringF[j] = CloudStringB[j] + CloudStringD +CloudStringE;', '', '', '\t}', '', "CloudStringG = CloudStringF.join(' ');", '', "CloudStringH = '      ]\t                                        
#' +", "'     }\t                                                        
#' +", "'#########################################################      
#';", '', 'CloudString = CloudStringA + CloudStringG + CloudStringH;', '', '', '', 'newNode = Browser.createVrmlFromString(CloudString);', 'Cirrus.children[i] = newNode[0];', '', '  }', '', '}', '', '', 'function initialize()', '', '{', '', 'cumulus();', '', 'cirrus();', '}', '', '']
),
DirectionalLight(ambientIntensity=1, color=((1,0,0)), direction=((-1,-1,0)), global_=True)]),]))
.XML())

File: src/main/dart/net/coderextreme/data/CloudsProcedural4.dart
// Copyright 2021 by John Carlson.  All rights reserved.
import 'dart:io';
import '../x3d.dart';
var X3D0 =  X3D(

      profile_ : SFString('Immersive'),
      version_ : SFString('3.2'),
      head_ : 
        head(
          meta_ : [
            meta(
              name_ : SFString('title'),
              content_ : SFString('CloudsProcedural4.x3d')),

            meta(
              name_ : SFString('description'),
              content_ : SFString('X3D utilizing ecmascript to develop quasi volumetric 3D clouds from png image textured billboard nodes.')),

            meta(
              name_ : SFString('creator'),
              content_ : SFString('Capt Darren W. Murphy')),

            meta(
              name_ : SFString('created'),
              content_ : SFString('1 November 2007')),

            meta(
              name_ : SFString('modified'),
              content_ : SFString('14 January 2014')),

            meta(
              name_ : SFString('identifier'),
              content_ : SFString('https://savage.nps.edu/Savage/Environment/Atmosphere/CloudsProcedural4.x3d')),

            meta(
              name_ : SFString('generator'),
              content_ : SFString('X3D-Edit, https://www.web3d.org/x3d/content/README.X3D-Edit.html')),

            meta(
              name_ : SFString('license'),
              content_ : SFString('../../license.html')),

            meta(
              name_ : SFString('TODO'),
              content_ : SFString('fix links'))]),
      Scene_ : 
        Scene(
          /*A png image file for the cloud texture must be designated in the ecmascript node.*/
          children_ : [
            Viewpoint(
              description_ : SFString('Main'),
              jump_ : false,
              orientation_ : SFRotation([SFDouble(0), SFDouble(1), SFDouble(0), SFDouble(1.57)]),
              position_ : SFVec3f([SFDouble(50000), SFDouble(1000), SFDouble(42000)])),

            Viewpoint(
              description_ : SFString('Light House Tower'),
              jump_ : false,
              orientation_ : SFRotation([SFDouble(0), SFDouble(1), SFDouble(0), SFDouble(1.3)]),
              position_ : SFVec3f([SFDouble(45000), SFDouble(1290), SFDouble(44000)])),

            Viewpoint(
              description_ : SFString('centerWest'),
              jump_ : false,
              orientation_ : SFRotation([SFDouble(0), SFDouble(1), SFDouble(0), SFDouble(2.5)]),
              position_ : SFVec3f([SFDouble(48000), SFDouble(1000), SFDouble(20000)])),

            Background(
              groundColor_ : MFColor([SFColor(0), SFColor(0), SFColor(1)]),
              skyColor_ : MFColor([SFColor(0), SFColor(0), SFColor(1)])),

            DirectionalLight(
              ambientIntensity_ : 1,
              direction_ : SFVec3f([SFDouble(-1), SFDouble(0), SFDouble(0)]),
              global_ : true),

            Group(
              DEF_ : SFString('Terrain'),
              children_ : [
                Transform(
                  scale_ : SFVec3f([SFDouble(50), SFDouble(50), SFDouble(50)]),
                  translation_ : SFVec3f([SFDouble(25000), SFDouble(0), SFDouble(25000)]),
                  children_ : [
                    Inline(
                      url_ : MFString([SFString("MontereyBayLargeMesh.x3d"), SFString("https://savage.nps.edu/Savage/Environment/Atmosphere/MontereyBayLargeMesh.x3d"), SFString("MontereyBayLargeMesh.wrl"), SFString("https://savage.nps.edu/Savage/Environment/Atmosphere/MontereyBayLargeMesh.wrl")]))]),

                Transform(
                  rotation_ : SFRotation([SFDouble(1), SFDouble(0), SFDouble(0), SFDouble(1.57)]),
                  translation_ : SFVec3f([SFDouble(25000), SFDouble(0), SFDouble(25000)]),
                  children_ : [
                    Shape(
                      geometry_ : 
                        Rectangle2D(
                          size_ : SFVec2f([SFDouble(77000), SFDouble(55000)])),
                      appearance_ : 
                        Appearance(
                          texture_ : 
                            ImageTexture(
                              url_ : MFString([SFString("ocean.png"), SFString("https://savage.nps.edu/Savage/Environment/Atmosphere/ocean.png")]))))])]),

            Group(
              DEF_ : SFString('Placemarks'),
              children_ : [
                Transform(
                  scale_ : SFVec3f([SFDouble(50), SFDouble(50), SFDouble(50)]),
                  translation_ : SFVec3f([SFDouble(45000), SFDouble(30), SFDouble(44000)]),
                  children_ : [
                    Inline(
                      url_ : MFString([SFString("Lighthouse.x3d"), SFString("https://savage.nps.edu/Savage/Environment/Atmosphere/Lighthouse.x3d"), SFString("Lighthouse.wrl"), SFString("https://savage.nps.edu/Savage/Environment/Atmosphere/Lighthouse.wrl")]))])]),

            Group(
              DEF_ : SFString('Clouds'),
              children_ : [
                Transform(
                  DEF_ : SFString('Cumulus')),

                Transform(
                  DEF_ : SFString('Cirrus')),

                Transform(
                  DEF_ : SFString('Fog')),

                Script(
                  DEF_ : SFString('PixelScript'),
                  directOutput_ : true,
                  field_ : [
                    field(
                      type_ : SFString("SFNode"),
                      name_ : SFString('Cumulus'),
                      accessType_ : SFString("initializeOnly"),
                      children_ : [
                        Transform(
                          USE_ : SFString('Cumulus'))]),

                    field(
                      type_ : SFString("SFNode"),
                      name_ : SFString('Cirrus'),
                      accessType_ : SFString("initializeOnly"),
                      children_ : [
                        Transform(
                          USE_ : SFString('Cirrus'))]),

                    field(
                      type_ : SFString("SFNode"),
                      name_ : SFString('Fog'),
                      accessType_ : SFString("initializeOnly")),
                  ],
ecmascript:eval (0
Cirrus.children = [];


, function cumulustranslation() // These values designate the boundary location of the cloud
{
	X = 50000*Math.random();          //  X horizontal range
	Y = 1000 + 300*Math.random();	 //  Y vertical base + range
	Z = 50000*Math.random();         // z horizontal range

	randomt = new String(X+' '+Y+' '+Z);

	return randomt;
	
}



, function cumulusscale() // these values scale a cloud within a designated size
{

	var maxscale = 1;

	scale = Math.round(9+maxscale*Math.random());
	X = 1.5*scale;
	Y = scale;
	Z = scale;

	randomscale = new String(X+' '+Y+' '+Z);

	return randomscale;
	
}


, function cirrustranslation() // These values designate the boundary location of the cloud
{
	X = 50000*Math.random();          //  X horizontal range
	Y = 8000 + 1000*Math.random();	 //  Y vertical base + range
	Z = 50000*Math.random();         // z horizontal range

	randomt = new String(X+' '+Y+' '+Z);

	return randomt;
	
}



, function cirrusscale() // these values scale a cloud within a designated size
{

	var maxscale = 1;

	scale = Math.round(9+maxscale*Math.random());
	X = 1.5*scale;
	Y = 2*Math.random();
	Z = 1.5*scale;

	randomscale = new String(X+' '+Y+' '+Z);

	return randomscale;
	
}


, function cumulussectiontranslation() // These random values place another portion of cumulus type cloud
{

	randomtheta = 6.28319*Math.random();
	randomphi = .7854*Math.random();
	randomradius = 90 + 5*Math.random();//the first whole number should be close to the sectionradius

	X = randomradius*Math.cos(randomtheta)*Math.sin(randomphi);
	Z = randomradius*Math.sin(randomtheta)*Math.sin(randomphi);
	Y = randomradius*Math.cos(randomphi);
	

	randomt = new String(X+' '+Y+' '+Z);

	return randomt;
	
}

, function cirrussectiontranslation() // These random values place another portion of cirrus type cloud
{

	randomtheta = 6.28319*Math.random();
	randomphi = .7854*Math.random();
	randomradius = 90 + 5*Math.random();//the first whole number should be close to the sectionradius

	X = randomradius*Math.cos(randomtheta)*Math.sin(randomphi);
	Z = randomradius*Math.sin(randomtheta)*Math.sin(randomphi);
	Y = randomradius*Math.cos(randomphi);
	

	randomt = new String(X+' '+Y+' '+Z);

	return randomt;
	
}


, function rotation() // This random value is for the billboard rotation not used in this script
{


	radians = 6.28*Math.random();

	randomr = new String('0 0 1 ' + radians );
	
	
	return randomr;
	
}

, function cumulus()
{

var maxi = 20;  // number of clouds

var maxj = 5; // denotes how many portions affecting the size of the cloud

var maxk = 8;  // number of billboards indicating cloud density

var sectionradius = 100;  //radius of individual cloud sections





for (var i=0; i < maxi; i++) 
{



CloudStringA = '	Transform {		\n' +
'    scale '+ cumulusscale() + '               	\n' +
'    translation '+ cumulustranslation() + '    \n' +    // cloud placement
'    children [	                                \n';


CloudStringB = new Array();
CloudStringF = new Array();

   	for (var j=0; j < maxj; j++)
   	{

	radius = 0;

	CloudStringB[j]= '  Transform {		    	       \n' +
	'    translation '+ cumulussectiontranslation() + '    \n' +     // section placement
	'    children [	                                       \n';

	
	CloudStringC = new Array();
	image = new String();

      		for (var k=1; k < maxk; k++)  // maxk value denotes how many textured billboards make up the cloud 
      		{


		randomtheta = 6.28319*Math.random();
		randomphi = 1.57079*Math.random();
		radius = radius+(sectionradius/maxk); // radius incremental steps based on billow radius and max billboards

		X = radius*Math.cos(randomtheta)*Math.sin(randomphi);
		Z = radius*Math.sin(randomtheta)*Math.sin(randomphi);
		Y = radius*Math.cos(randomphi);


		if (Y <= 30) //cloud shading and lighting control
  	{	
	image = ' "CloudTexture1_5.png" "https://savage.nps.edu/Savage/Environment/Spheretexture.png" \n';
  	}

  		else
  	{	
	image = ' "CloudTexture1_4.png" "https://savage.nps.edu/Savage/Environment/Spheretexture.png" \n';
  	}

	
		
		Billboardtranslation = new String(X+' '+Y+' '+Z);

		CloudStringC[k] = '	Transform {		                \n' +
		'            translation '+ Billboardtranslation   + '          \n' +     // random billboard placement within radius designated above
		'	  children [	                                        \n' +
		'	      Billboard {	                                \n' +
		'	        axisOfRotation 0 0 0	                        \n' +     // 0 0 0 designates rotation on all axis
		'	        children [	                                \n' +
		'	            Transform {	                		\n' +
		'	              rotation  0 0 0 0 		        \n' +     // a rotation of the individual billboards can be defined
		'	              children [	                        \n' +
		'	                  Shape {	                        \n' +
		'	                    appearance Appearance {	        \n' +
		'				material Material {		\n' +
		'				                }  		\n' +
		'	                      texture ImageTexture {	        \n' +
		'	                        url [ ' + image + ' ]           \n' + 
		'	                      }	                                \n' +
		'	                    }	                                \n' +
		'	                    geometry IndexedFaceSet {	        \n' +     // define type of geometry to texture
		'	                      coordIndex [ 0, 1, 2, 3 ]	        \n' +
		'			      solid FALSE		        \n' +
		'	                      coord Coordinate {	        \n' +
		'	                        point [ 50 50 0,	        \n' +     // define size of the geometry. Here 100 meter 2D square.
		'	                                50 -50 0,	        \n' +
		'	                               -50 -50 0,	        \n' +
		'	                               -50 50 0 ]	        \n' +
		'	                      }	                                \n' +
		'	                    }	                                \n' +
		'	                  }	                                \n' +
		'	              ]	                                        \n' +
		'	            }	                                        \n' +
		'	       ]	                                        \n' +
		'	   }	                                                \n' +
		'      ]	                                                \n' +
		'     }	                                                        \n';      
		

		}

	CloudStringD = CloudStringC.join(' ');

	
	CloudStringE = '   ]	                 \n' +
	'	}	                         \n';

	CloudStringF[j] = CloudStringB[j] + CloudStringD +CloudStringE;


	}

CloudStringG = CloudStringF.join(' ');

CloudStringH = '      ]	                                        \n' +
'     }	                                                        \n' +
'#########################################################      \n';

CloudString = CloudStringA + CloudStringG + CloudStringH;



newNode = Browser.createVrmlFromString(CloudString);
Cumulus.children[i] = newNode[0];


   }

}

, function cirrus()

{

var maxi = 2;  // number of clouds

var maxj = 5; // denotes how many portions affecting the size of the cloud

var maxk = 8;  // number of billboards indicating cloud density

var sectionradius = 1000;  //radius of individual cloud sections





for (var i=0; i < maxi; i++) 
{



CloudStringA = '	Transform {		 \n' +
'    scale '+ cirrusscale() + '               	 \n' +
'    translation '+ cirrustranslation() + '      \n' +    // cloud placement
'    children [	                                 \n';


CloudStringB = new Array();
CloudStringF = new Array();

   	for (var j=0; j < maxj; j++)
   	{

	radius = 0;

	CloudStringB[j]= '  Transform {		    	      \n' +
	'    translation '+ cirrussectiontranslation() + '    \n' +     // section placement
	'    children [	                                      \n';

	
	CloudStringC = new Array();

      		for (var k=1; k < maxk; k++)  // maxk value denotes how many textured billboards make up the cloud 
      		{


		randomtheta = 6.28319*Math.random();
		randomphi = 1.57079*Math.random();
		radius = radius+(sectionradius/maxk); // radius incremental steps based on section radius and max billboards

		X = radius*Math.cos(randomtheta)*Math.sin(randomphi);
		Z = radius*Math.sin(randomtheta)*Math.sin(randomphi);
		Y = radius*Math.cos(randomphi);
		
		Billboardtranslation = new String(X+' '+Y+' '+Z);

		CloudStringC[k] = '	Transform {		                \n' +
		'            translation '+ Billboardtranslation   + '          \n' +     // random billboard placement within radius designated above
		'	  children [	                                        \n' +
		'	      Billboard {	                                \n' +
		'	        axisOfRotation 0 0 0	                        \n' +     // 0 0 0 designates rotation on all axis
		'	        children [	                                \n' +
		'	            Transform {	                		\n' +
		'	              rotation '  + rotation() + '	        \n' +
		'	              children [	                        \n' +
		'	                  Shape {	                        \n' +
		'	                    appearance Appearance {	        \n' +
		'			    material Material {			\n' +
		'			    }					\n' +
 		'	                      texture ImageTexture {	        \n' +
		'	                        url ["cloudtexture3.png" "https://savage.nps.edu/Savage/Environment/cloudtexture1_4.png" ] \n' +
		'	                      }	                                \n' +
		'	                    }	                                \n' +
		'	                    geometry IndexedFaceSet {	        \n' +     // define type of geometry to texture
		'	                      coordIndex [ 0, 1, 2, 3 ]	        \n' +
		'			      solid FALSE		        \n' +
		'	                      coord Coordinate {	        \n' +
		'	                        point [ 500 500 0,	        \n' +     // define size of the geometry. Here 100 meter 2D square.
		'	                                500 -500 0,	        \n' +
		'	                               -500 -500 0,	        \n' +
		'	                               -500 500 0 ]	        \n' +
		'	                      }	                                \n' +
		'	                    }	                                \n' +
		'	                  }	                                \n' +
		'	              ]	                                        \n' +
		'	            }	                                        \n' +
		'	       ]	                                        \n' +
		'	   }	                                                \n' +
		'      ]	                                                \n' +
		'     }	                                                        \n';      
		

		}

	CloudStringD = CloudStringC.join(' ');

	CloudStringE = '   ]	                 \n' +
	'	}	                         \n';

	CloudStringF[j] = CloudStringB[j] + CloudStringD +CloudStringE;


	}

CloudStringG = CloudStringF.join(' ');

CloudStringH = '      ]	                                        \n' +
'     }	                                                        \n' +
'#########################################################      \n';

CloudString = CloudStringA + CloudStringG + CloudStringH;



newNode = Browser.createVrmlFromString(CloudString);
Cirrus.children[i] = newNode[0];

  }

}


, function initialize()

{

cumulus();

cirrus();
})),

                DirectionalLight(
                  ambientIntensity_ : 1,
                  color_ : SFColor([SFDouble(1), SFDouble(0), SFDouble(0)]),
                  direction_ : SFVec3f([SFDouble(-1), SFDouble(-1), SFDouble(0)]),
                  global_ : true)])]));
void main() { exit(0); }
